src/widgets/index/__tests__/index-test.ts

```diff
@@ -0,0 +1,871 @@
+      ...args,
+    });
+
+  it('throws without argument', () => {
+    expect(() => {
+      // @ts-ignore
+      index();
+    }).toThrowErrorMatchingInlineSnapshot(`
+"The \`indexName\` option is required.
+
+See documentation: https://www.algolia.com/doc/api-reference/widgets/index/js/"
+`);
+  });
+
+  it('throws without `indexName` option', () => {
+    expect(() => {
+      index({} as any);
+    }).toThrowErrorMatchingInlineSnapshot(`
+"The \`indexName\` option is required.
+
+See documentation: https://www.algolia.com/doc/api-reference/widgets/index/js/"
+`);
+  });
+
+  describe('addWidgets', () => {
+    it('adds given widgets to the instance', () => {
+      const instance = index({ indexName: 'index_name' });
+
+      expect(instance.getWidgets()).toHaveLength(0);
+
+      instance.addWidgets([createSearchBox(), createPagination()]);
+
+      expect(instance.getWidgets()).toHaveLength(2);
+    });
+
+    it('returns the instance to be able to chain the calls', () => {
+      const topLevelIndex = index({ indexName: 'top_level_index_name' });
+      const subLevelIndex = index({ indexName: 'sub_level_index_name' });
+
+      topLevelIndex.addWidgets([
+        subLevelIndex.addWidgets([createSearchBox(), createPagination()]),
+      ]);
+
+      expect(topLevelIndex.getWidgets()).toHaveLength(1);
+      expect(topLevelIndex.getWidgets()).toEqual([subLevelIndex]);
+    });
+
+    it('does not throw an error without the `init` step', () => {
+      const topLevelIndex = index({ indexName: 'top_level_index_name' });
+      const subLevelIndex = index({ indexName: 'sub_level_index_name' });
+
+      expect(() => topLevelIndex.addWidgets([subLevelIndex])).not.toThrow();
+    });
+
+    it('throws an error with a value different than `array`', () => {
+      const instance = index({ indexName: 'index_name' });
+
+      expect(() => {
+        // @ts-ignore
+        instance.addWidgets();
+      }).toThrow();
+
+      expect(() => {
+        instance.addWidgets(createWidget() as any);
+      }).toThrowErrorMatchingInlineSnapshot(`
+"The \`addWidgets\` method expects an array of widgets.
+
+See documentation: https://www.algolia.com/doc/api-reference/widgets/index/js/"
+`);
+    });
+
+    it('throws an error with widgets that do not implement `init` or `render`', () => {
+      const instance = index({ indexName: 'index_name' });
+
+      expect(() => {
+        instance.addWidgets([{ dummy: true } as any]);
+      }).toThrowErrorMatchingInlineSnapshot(`
+"The widget definition expects a \`render\` and/or an \`init\` method.
+
+See documentation: https://www.algolia.com/doc/api-reference/widgets/index/js/"
+`);
+    });
+
+    describe('with a started instance', () => {
+      it('updates the internal state with added widgets', () => {
+        const instance = index({ indexName: 'index_name' });
+        const instantSearchInstance = createInstantSearch();
+
+        instance.addWidgets([createSearchBox()]);
+
+        instance.init!(
```

Comment:
IIRC, `!` is necessary here because `Index = {...} & Widget` and `Widget` has optional `init` method.
I wonder if it wouldn't have been preferable to enforce `init` (and `render`) on Index (AFAIKT they are always there).
Not sure it brings a lot of value but you like to hear your opinion on this.

Reply:
Yes we can enforce it. Like you said it doesn't bring that much value though. The `init`/`render` methods are only (almost) called directly inside the tests. Most of the time the widgets are inside an array and we have to check whether or not the methods exist before calling them.

Reply:
https://github.com/algolia/instantsearch.js/pull/3892/commits/e4e4152031388cf07e5984bc129c83abd96d440b